### PR TITLE
Introduce IOContext.LOAD

### PR DESCRIPTION
**Pull request by jpountz**
This PR introduces a new `LOAD` `IOContext` for files that are a small fraction of the total index size and are expected to be accessed in a random-access fashion. This may be leveraged by some `Directory` implementations to load such files into physical memory (e.g. Java heap) in order to provide stronger guarantees on query latency.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/store/IOContext.java [hunk 1]
```diff
@@ -41,34 +41,47 @@ public enum Context {
...
+  /** This flag indicates that the file will be opened, then fully read sequentially then closed. */
   public final boolean readOnce;
 
+  /**
+   * This flag is used for files that are a small fraction of the total index size and are expected
+   * to be heavily accessed in random-access fashion. Some {@link Directory} implementations may
+   * choose to load such files into physical memory (e.g. Java heap) as a way to provide stronger
+   * guarantees on query latency.
+   */
+  public final boolean load;
```
Comment (uschindler) [inline lucene/core/src/java/org/apache/lucene/store/IOContext.java:53]: This description looks good.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90DocValuesProducer.java [hunk 2]
```diff
@@ -82,7 +83,7 @@ final class Lucene90DocValuesProducer extends DocValuesProducer {
     merging = false;
 
     // read in the entries from the metadata file.
-    try (ChecksumIndexInput in = state.directory.openChecksumInput(metaName, state.context)) {
+    try (ChecksumIndexInput in = state.directory.openChecksumInput(metaName, IOContext.READONCE)) {
```
Comment (uschindler) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90DocValuesProducer.java:86]: I was wondering about this earlier, but it is really nice you fixed this,too!

**Review by uschindler — APPROVED** [review]
Looks fine.

Maybe we should still add the the `BiPredicate<String,IOContext>` constant to MMapDirectory in this PR (but not eanble it by default). This would require that #11929 is merged first.

**Review by uschindler — APPROVED** [review]

**Comment by uschindler** [comment]
As we started over here: My suggestion would be to let this go in without changing the default in MMapDirectory.

If we want to change the default we can make a followup PR on MMapDirectory. But at moment due to disagreement if preload is enough, we should just add the option to enable this in MMapDir. If we really want to stick the small files and load them to heap, maybe a NRTCachingDirectory like approach would be best.

**Comment by jpountz** [comment]
Agreed Uwe.

**Comment by uschindler** [comment]
I think a CHANGES.txt entry would be good.

**Pull request merged**